tests/Unit/API/Site/Controllers/CountryCodeTraitTest.php

```diff
@@ -0,0 +1,248 @@
+	use Automattic\WooCommerce\GoogleListingsAndAds\API\Site\Controllers\CountryCodeTrait;
+	use Automattic\WooCommerce\GoogleListingsAndAds\Google\GoogleHelper;
+	use Automattic\WooCommerce\GoogleListingsAndAds\Vendor\League\ISO3166\ISO3166DataProvider;
+	use PHPUnit\Framework\MockObject\MockObject;
+	use PHPUnit\Framework\TestCase;
+	use Exception;
+	use WP_Error;
+	use WP_REST_Request as Request;
+
+	/**
+	 * Class CountryCodeTraitTest
+	 *
+	 * @package Automattic\WooCommerce\GoogleListingsAndAds\Tests\Unit\API\Site\Controllers
+	 * @phpcs:disable Squiz.Classes.ClassDeclaration.SpaceBeforeKeyword
+	 * @phpcs:disable Squiz.Classes.ClassDeclaration.SpaceBeforeCloseBrace
+	 */
+	class CountryCodeTraitTest extends TestCase {
+
+		/** @var callable $rest_validate_request_arg_callback */
+		public static $rest_validate_request_arg_callback = null;
+
+		/** @var MockObject|ISO3166DataProvider $iso_provider */
+		protected $iso_provider;
+
+		/** @var MockObject|GoogleHelper $google_helper */
+		protected $google_helper;
+
+		/** @var bool $country_supported */
+		protected $country_supported;
+
+		/**
+		 * Runs before each test is executed.
+		 */
+		public function setUp(): void {
+			parent::setUp();
+
+			$this->iso_provider  = $this->createMock( ISO3166DataProvider::class );
+			$this->google_helper = $this->createMock( GoogleHelper::class );
+
+			// phpcs:ignore WordPress.Classes.ClassInstantiation.MissingParenthesis
+			$this->trait = new class {
+				use CountryCodeTrait {
+					get_country_code_sanitize_callback as public;
+					get_country_code_validate_callback as public;
+					get_supported_country_code_validate_callback as public;
+				}
+			};
+
+			$this->trait->set_iso3166_provider( $this->iso_provider );
+			$this->trait->set_google_helper_object( $this->google_helper );
+
+			$this->country_supported = true;
+
+			$this->google_helper->method( 'is_country_supported' )
+				->willReturnCallback(
+					function () {
+						return $this->country_supported;
+					}
+				);
+		}
+
+		/**
+		 * Runs after each test is executed.
+		 */
+		public function tearDown(): void {
+			self::$rest_validate_request_arg_callback = null;
+		}
+
+		/**
+		 * Test all get_*_callback methods return a callable.
+		 */
+		public function test_all_get_callbacks_return_callable() {
+			$this->assertIsCallable( $this->trait->get_country_code_sanitize_callback() );
+			$this->assertIsCallable( $this->trait->get_country_code_validate_callback() );
+			$this->assertIsCallable( $this->trait->get_supported_country_code_validate_callback() );
+		}
+
+		/**
+		 * Test the get_country_code_validate_callback and get_supported_country_code_validate_callback
+		 * methods will apply arguments to rest_validate_request_arg_callback and return early if invalid.
+		 */
+		public function test_all_get_validate_callbacks_with_rest_validate_request_arg() {
+			$called_count = 0;
+
+			self::$rest_validate_request_arg_callback = function () use ( &$called_count ) {
```

Comment:
> I'm unsure how beneficial it is to replace the global `rest_validate_request_arg` so that we can monitor how many times it's called here.

This testing goal is to verify that the two callbacks do call the global `rest_validate_request_arg` to further ensure that the built-in schema validation of WP REST API is applied.

It's indeed very complicated and I got stuck with this part the longest. But PHPUnit seems not to provide an explicit way to mock a global function, and I couldn't find a better way to achieve this goal, so I ended up with the current solution.

> It makes it harder to understand and feels quite heavy-handed for what we're trying to test. Had you considered adding a wrapper for `rest_validate_request_arg` to `CountryCodeTrait` instead?
> 
> If we were using a wrapper then when setting up `$this->trait` in the test we could modify how that functions so that we can set the desired return values by adding something like: [...]
> This would avoid the need for two namespace declarations in the file and we wouldn't need to replace the global function.

Sorry, I can't understand from the provided code example how this approach is going to achieve the goals mentioned above. I tried to change some tests with it, but it didn't even be able to mock the result of calling `rest_validate_request_arg`.

Could you explain the solution a bit more?

Reply:
Thank you for the idea.

On the balance of robustness, complexity and WP coding standards, I agree it's feasible to indirectly verify "Whether the global `rest_validate_request_arg` is called."

Since the main purpose is not to mock the call result of `rest_validate_request_arg`, I omitted the part about creating the wrapper method in `CountryCodeTrait`.

I adjusted it by two commits:
1. https://github.com/woocommerce/google-listings-and-ads/pull/2004/commits/c6f320fc05e3b75707f012dece2c73499aa7b306 for the main change
2. https://github.com/woocommerce/google-listings-and-ads/pull/2004/commits/18c7d9599fe5ba7510cb7f35a03634e9358b8a2b for indentation adjustment

I believe it would be easier to review the changes.